### ISSUE TITLE
Rendering problems with FigureCanvasWxAgg on OSX

### ISSUE THREAD

**Issue by mattsbjorck** [matplotlib/matplotlib#10174]
### Bug report

**Bug summary**

Including FigureCanvasWxAgg in wx app with log scale axis causes issues with repainting the window, sometime the Figure is painted in the wrong place or another panel is painted over it. 
Removing the log scale (using a linear scale) causes it to paint correctly. This work in older versions of wxpython(3.0.0) and python(2.7.6). issue appears on OSX.

**Code for reproduction**

```python
"""Script to test and explore the issue with bad repainting of a the SplitterWindow."""
import wx

import matplotlib as mpl
from matplotlib.backends.backend_wxagg import FigureCanvasWxAgg as Canvas


class Plot(wx.Panel):
    def __init__(self, parent, id=-1, dpi=None, **kwargs):
        wx.Panel.__init__(self, parent, id=id, **kwargs)
        self.figure = mpl.figure.Figure(dpi=dpi, figsize=(2, 2))
        self.canvas = Canvas(self, -1, self.figure)

        sizer = wx.BoxSizer(wx.VERTICAL)
        sizer.Add(self.canvas, 1, wx.EXPAND)
        self.SetSizer(sizer)


class MainFrame(wx.Frame):
    def __init__(self, *args, **kwds):
        wx.Frame.__init__(self, *args, **kwds)

        self.ver_splitter = wx.SplitterWindow(self, -1, style=wx.SP_3D)
        self.hor_splitter = wx.SplitterWindow(self.ver_splitter, -1, style=wx.SP_3D)

        self.input_notebook = wx.Notebook(self.hor_splitter, -1, style=wx.NB_BOTTOM | wx.NO_BORDER)
        self.plot_notebook = wx.Notebook(self.hor_splitter, -1, style=wx.NB_BOTTOM | wx.NO_BORDER)

        self.panel_left = wx.Notebook(self.ver_splitter)

        self.plot_data = Plot(self.plot_notebook)
        self.plot_notebook.AddPage(self.plot_data, "Data")


        self.ver_splitter.SplitVertically(self.panel_left, self.hor_splitter)
        self.hor_splitter.SplitHorizontally(self.plot_notebook, self.input_notebook)

        self.Layout()

        axes = self.plot_data.figure.gca()
        axes.plot([1, 2, 3], [1, 10, 100.])
       # Comment out the line below to remove the problem.
        axes.set_yscale('log')

if __name__ == "__main__":
    app = wx.App(False)
    main_frame = MainFrame(None)
    main_frame.Show()
    app.MainLoop()
```

**Actual outcome**

See screenshot below. This is not static but appears sometimes upon resizing the window or the sizers. Sometimes the plot is plotted in the wrong place as well. The gray part of the plot is actually the panel located on the left painted on the wrong place. 

<img width="822" alt="issue1" src="https://user-images.githubusercontent.com/8424827/34619937-1d8e5fc0-f244-11e7-9ce3-7f08bfde492c.png">

**Matplotlib version**
Mac OSX 10.12.6
Python 3.6.1
wxpython 4.0.0b2 osx-cocoa (phoenix)
Matplotlib 2.1.1

Python installed from miniconda, wxpython with pip.

Any help or pointers that could help are welcome!

Matts Björck

**Comment by jklymak** [comment]
Can you clarify: it works with Matplotlin 2.1.1, py 2.7, wxpython 3.0.0?  Or did you also change Matplotlib versions?

**Comment by mattsbjorck** [comment]
The working version is with matplotlib 1.4.3, will update it and try again. So matplotlib versions have changed as well.

**Comment by jklymak** [comment]
Looks even worse for me.  Is that the exact code you are using?

<img width="432" alt="4__pythonw_testwx_py__python_" src="https://user-images.githubusercontent.com/1562854/34621336-9f340968-f1fd-11e7-89dc-fab355004517.png">

**Comment by jklymak** [comment]
If I resize the windows all sorts of funkiness ensues;  note the left-over figure below the newly re-sized panel.  

<img width="662" alt="rendering_problems_with_figurecanvaswxagg_on_osx_ _issue__10174_ _matplotlib_matplotlib_and_4__pythonw_testwx_py__python_" src="https://user-images.githubusercontent.com/1562854/34621395-ccca2376-f1fd-11e7-94bc-a073977732a4.png">

**Comment by mattsbjorck** [comment]
Yes that is exactly the issues that I have. Upon resizing the window is not redrawn correctly. The code that I pasted is exactly the code that yields the problem. The original issue arose when porting an application of mine, GenX [http://genx.sf.net](http://genx.sf.net), to python 3 and then also used used updated libraries. So the code is a very minimal version that reproduce the problem of the real application.

By the way does the problems associated with resizing disappear if you remove the statement `axes.set_yscale('log')` ?

**Comment by jklymak** [comment]
Nope, just for log.  That does indicate an issue in Matplotlib.  But I'm not an expert on embedding matplotlib, so hopefully someone who does that sort of thing will have a look.

**Comment by mattsbjorck** [comment]
I have done some more testings. Everything works fine on Python2.7.4 with matplotlib 2.1.1 and wxPython 3.0 as well as 4.0.0b2.
Problems occur on Python 3.5.4 and matplotlib 2.1.1 and wxPython 4.0.0b2. 

I hope this might help you - I paste the configurations of the testing environments below.

## Python 2.7.4 condo list output - works
```
backports                 1.0                      py27_1    conda-forge
backports.functools_lru_cache 1.4                      py27_1    conda-forge
backports_abc             0.5                      py27_0    conda-forge
blas                      1.1                    openblas    conda-forge
ca-certificates           2017.11.5                     0    conda-forge
certifi                   2017.11.5                py27_0    conda-forge
cycler                    0.10.0                   py27_0    conda-forge
freetype                  2.8.1                         0    conda-forge
functools32               3.2.3.2                  py27_1    conda-forge
libgfortran               3.0.0                         0    conda-forge
libpng                    1.6.34                        0    conda-forge
matplotlib                2.1.1                    py27_2    conda-forge
ncurses                   5.9                          10    conda-forge
numpy                     1.9.3           py27_blas_openblas_203  [blas_openblas]  conda-forge
openblas                  0.2.20                        6    conda-forge
openssl                   1.0.2n                        0    conda-forge
pip                       9.0.1                    py27_1    conda-forge
pyparsing                 2.0.3                    py27_0  
pyqt                      4.11.4                   py27_2    conda-forge
python                    2.7.14                        4    conda-forge
python-dateutil           2.6.1                    py27_0    conda-forge
python.app                1.2                      py27_0    conda-forge
pytz                      2017.3                     py_2    conda-forge
qt                        4.8.7                         4  
readline                  7.0                           0    conda-forge
scipy                     1.0.0           py27_blas_openblas_201  [blas_openblas]  conda-forge
setuptools                38.2.4                   py27_0    conda-forge
singledispatch            3.4.0.3                  py27_0    conda-forge
sip                       4.18                     py27_1    conda-forge
six                       1.11.0                   py27_1    conda-forge
sqlite                    3.20.1                        2    conda-forge
ssl_match_hostname        3.5.0.1                  py27_1    conda-forge
subprocess32              3.2.7                    py27_0    conda-forge
tk                        8.6.7                         0    conda-forge
tornado                   4.5.2                    py27_0    conda-forge
wheel                     0.30.0                     py_1    conda-forge
wxPython                  4.0.0b2                   <pip>
zlib                      1.2.11                        0    conda-forge
```

## Python 3.5.4 conda list -issues
```
backports                 1.0                      py35_1    conda-forge
backports.functools_lru_cache 1.4                      py35_1    conda-forge
ca-certificates           2017.11.5                     0    conda-forge
certifi                   2017.11.5                py35_0    conda-forge
cycler                    0.10.0                   py35_0    conda-forge
freetype                  2.8.1                         0    conda-forge
libpng                    1.6.34                        0    conda-forge
matplotlib                2.1.1                    py35_2    conda-forge
mkl                       11.3.3                        0  
ncurses                   5.9                          10    conda-forge
numpy                     1.10.4                   py35_2  
openssl                   1.0.2n                        0    conda-forge
pip                       9.0.1                    py35_1    conda-forge
pyparsing                 2.0.3                    py35_0  
pyqt                      4.11.4                   py35_2    conda-forge
python                    3.5.4                         3    conda-forge
python-dateutil           2.6.1                    py35_0    conda-forge
python.app                1.2                      py35_0    conda-forge
pytz                      2017.3                     py_2    conda-forge
qt                        4.8.7                         4  
readline                  7.0                           0    conda-forge
setuptools                38.2.4                   py35_0    conda-forge
sip                       4.18                     py35_1    conda-forge
six                       1.11.0                   py35_1    conda-forge
sqlite                    3.20.1                        2    conda-forge
tk                        8.6.7                         0    conda-forge
tornado                   4.5.2                    py35_0    conda-forge
wheel                     0.30.0                     py_1    conda-forge
wxPython                  4.0.0b2                   <pip>
xz                        5.2.3                         0    conda-forge
zlib                      1.2.11                        0    conda-forge
```

**Comment by WeatherGod** [comment]
I find it really strange that this would be dependent upon log scale +
embedding in wx. Now, I do know that the wx backend in the past has needed
some extra draw() (as opposed to draw_idle()) calls. In my book, I
suggested the phrase "kick in the pants", but my editor rejected that. I
never tracked down the cause for this, though.

**Comment by DietmarSchwertberger** [comment]
@mattsbjorck: could you please try with `evt.Skip()` commented out? With `evt.Skip()` I see similar problems on Windows.
It's in `matplotlib\backends\backend_wx.py`, class `FigureCanvasWx`
```
    def _onPaint(self, evt):
        drawDC = wx.PaintDC(self)
        if not self._isDrawn:
            self.draw(drawDC=drawDC)
        else:
            self.gui_repaint(drawDC=drawDC)
        #evt.Skip()
```

**Comment by mattsbjorck** [comment]
@DietmarSchwertberger, I have currently no source installation of matplotlib setup. So I subclassed FigureCanvasWxAgg according to the following code:
```
class CanvasTest(Canvas):
    def _onPaint(self, evt):
        drawDC = wx.PaintDC(self)
        if not self._isDrawn:
            self.draw(drawDC=drawDC)
        else:
            self.gui_repaint(drawDC=drawDC)
         #evt.Skip()


class Plot(wx.Panel):
    def __init__(self, parent, id=-1, dpi=None, **kwargs):
        wx.Panel.__init__(self, parent, id=id, **kwargs)
        self.figure = mpl.figure.Figure(dpi=dpi, figsize=(2, 2))
        self.canvas = CanvasTest(self, -1, self.figure)

        sizer = wx.BoxSizer(wx.VERTICAL)
        sizer.Add(self.canvas, 1, wx.EXPAND)
        self.SetSizer(sizer)
```
which I believe should give the same effect. The problem remains with `evt.Skip()` commented out. 

However, exchanging `FigureCanvasWxAgg` to `FigureCanvasWx` in my original post will cause the application to render correctly. I don't know if that gives any new information.

**Comment by WeatherGod** [comment]
Interesting. Yeah, in that case the evt.Skip() would have no bearing on
this (since both FigureCanvasWx and FigureCanvasWxAgg are calling it one
way or another). This does let us focus our attention on the WxAgg version,
though.

**Comment by WeatherGod** [comment]
I think I see a bug. The blit() method in FigureCanvasWxAgg doesn't specify
the origin of the gui_repaint() calls. At the very least, that could cause
problems on some systems for blitting, which I think is how window resizes
are done.

**Comment by WeatherGod** [comment]
Interesting...

Wx's draw()
```
    def draw(self, drawDC=None):
        """
        Render the figure using RendererWx instance renderer, or using a
        previously defined renderer if none is specified.
        """
        DEBUG_MSG("draw()", 1, self)
        self.renderer = RendererWx(self.bitmap, self.figure.dpi)
        self.figure.draw(self.renderer)
        self._isDrawn = True
        self.gui_repaint(drawDC=drawDC)

    def gui_repaint(self, drawDC=None, origin='WX'):
        """
        Performs update of the displayed image on the GUI canvas, using the
        supplied wx.PaintDC device context.

        The 'WXAgg' backend sets origin accordingly.
        """
        DEBUG_MSG("gui_repaint()", 1, self)
        if self.IsShownOnScreen():
            if not drawDC:
                # not called from OnPaint use a ClientDC
                drawDC = wx.ClientDC(self)

            # following is for 'WX' backend on Windows
            # the bitmap can not be in use by another DC,
            # see GraphicsContextWx._cache
            if wx.Platform == '__WXMSW__' and origin == 'WX':
                img = self.bitmap.ConvertToImage()
                bmp = img.ConvertToBitmap()
                drawDC.DrawBitmap(bmp, 0, 0)
            else:
                drawDC.DrawBitmap(self.bitmap, 0, 0)

```

WxAgg's draw()

```
    def draw(self, drawDC=None):
        """
        Render the figure using agg.
        """
        DEBUG_MSG("draw()", 1, self)
        FigureCanvasAgg.draw(self)

        self.bitmap = _convert_agg_to_wx_bitmap(self.get_renderer(), None)
        self._isDrawn = True
        self.gui_repaint(drawDC=drawDC, origin='WXAgg')
```

So, WxAgg's draw() doesn't go through any renderer, at least, not
explicitly?

**Comment by DietmarSchwertberger** [comment]
With this code in `FigureCanvasWxAgg`  I get a black canvas, but the resizing is OK:
```
        #FigureCanvasAgg.draw(self)
        #self.bitmap = _convert_agg_to_wx_bitmap(self.get_renderer(), None)
        self.bitmap = _convert_agg_to_wx_bitmap(self.get_renderer(), None)
```
When I enable `FigureCanvasAgg.draw(self)` again, I get the repaint problems on resizing. Somehow this call seems to interfere with the repaint.

**Comment by DietmarSchwertberger** [comment]
Further down:
When I comment out these lines in `Figure.draw`, the resizing problems are gone:
```
            mimage._draw_list_compositing_images(
                renderer, self, artists, self.suppressComposite)
```
Of course, the contents of the canvas are also gone...

**Comment by DietmarSchwertberger** [comment]
With `_AxesBase` in `_base.py` modified:
```
mimage._draw_list_compositing_images(renderer, self, artists[1:2])
```
Now only the logarithmic y axis is drawn as it is the second element in `artists`. This seems to trigger the problem. When I select `[:1]` or `[2:]` then there are no resizing problems...

**Comment by mattsbjorck** [comment]
@DietmarSchwertberger, Yes I get the same type of problem with linear axes and math text. 
`axes.set_yticklabels(['$100$', '$10$', '$20$'])` yields buggy repainting while `axes.set_yticklabels(['100', '10', '20']) ` repaint the application properly.

**Comment by DietmarSchwertberger** [comment]
Unfortunately I'm at a dead end now.
In `mathtext.MathTextParser` there's a call to a `matplotlib.mathtext.Parser` instance:
```box = self._parser.parse(s, font_output, fontsize, dpi)```
If I avoid this call and return the previous result instead, then of course the display is not correct, but at least the resizing problems are gone.

This is difficult to debug as there a recursive pyparsing parser is entered. If I limit the parse depth, then the problems are also fixed.

So the question is: how is the mathtext parser linked to it's backend? I guess that somewhere there is a call to some wx function or a wx device or graphics context is created or something like that.
I'm not familiar enough with matplotlib to answer this.

**Comment by WeatherGod** [comment]
You are right... that doesn't seem to make any sense whatsoever. How are
you limiting the parse depth?

Hypothesis: what if pyparsing is actually hitting a recursion depth limit
and raising an exception that is getting accidentally caught by wx?

**Comment by DietmarSchwertberger** [comment]
When going through, I added an argument `return_early` to the relevant functions and did initialize it with e.g. `return_early=4`.
Some methods would not execute through to `return` with this argument >0 and just return a dummy or cached result. The recursive function would reduce it by one and return half-way if the value is below a threshold.
I ensured that only a simple string like `'$1$'` was to be parsed and my intial value for `return_early` was between 3 and 4 and the threshold 1 or 2. So it does not look like a deep recursion.


This might be unrelated, but sometimes I observed an error message `...Python[413] <Error>: CGContextRestoreGState: invalid context 0x0. Backtrace:`, which would be expanded if `os.environ["CG_CONTEXT_SHOW_BACKTRACE"]` is set to anything.
See attachment.
[backtrace.txt](https://github.com/matplotlib/matplotlib/files/1624146/backtrace.txt)

**Comment by WeatherGod** [comment]
New theory... a race condition is happening between successive redraw calls
triggered by resizing events. One of those events is probably invalidating
the graphics context, while another instance is trying to utilize the now
invalidated graphics context. Notice the memory location for the CG context
is 0x0, which is the Null pointer.

**Comment by DietmarSchwertberger** [comment]
Right to the toplevel of the backtrace, there's something like `dealloc` and `release_wxPaintDC`.

I just tried:
```
import gc
gc.disable()
```
This fixes the most issues for me, but is of course not suitable for production code. As containment, the gc could be disabled in the paint event handler.

@mattsbjorck: could you please check whether `gc.disable()` improves the situation for you?
I see only one repainting issue now: the shadow of the left panel is drawn on the matplotlib canvas.

What I don't like in the wx backends is the caching of graphics and device contexts. This could contribute to the problem. But then, the caching is mostly done for the non-Agg backend...
If it is confirmed that `gc.disable()` fixes the problem, then we should address this to Robin. Maybe he can check the backtrace.

**Comment by DietmarSchwertberger** [comment]
I think I've found a solution:
Add a `drawDC.Destroy()` to the end of `backend_wx.FigureCanvasWx._onPaint`.
An explicit call to `PaintDC.Destroy` usually should not be necessary. This might be a bug in wxPython. I will submit a report. The backtrace might be  a side effect when the DC is destroyed too late only when it's being collected by the garbage collector.

**Comment by tacaswell** [comment]
Closed by #10352
